### PR TITLE
c8y configuration plugin implementation

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

Implementing `fs_notify` crate for c8y configuration plugin. 

This commit contains the changes of #1244 applied on top of it. When reviewing please review only the last commit: 8e5f1f84f440e1f589cad6117b02be34540918e3

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
- #1248

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: plugins/c8y_configuration_plugin/src/upload.rs [hunk 1]
```diff
@@ -89,6 +89,18 @@ pub async fn handle_config_upload_request(
...
+    mqtt_client: &mut Connection,
+    config_file_path: &Path,
+) -> Result<PluginConfig> {
+    // Reload the plugin config file
+    let plugin_config = PluginConfig::new(config_file_path);
+    // Resend the supported config types
+    let msg = plugin_config.to_supported_config_types_message()?;
+    mqtt_client.published.send(msg).await?;
+    Ok(plugin_config)
+}
```
Comment (didier-wenzek) [inline plugins/c8y_configuration_plugin/src/upload.rs:102]: This function does too many things: reading the configuration and sending a message to c8y. I would do just the later given a `PluginConfig` and not a path to that config.

```suggestion
pub(crate) async fn handle_dynamic_config_update(
    mqtt_client: &mut Connection,
    plugin_config: &PluginConfig,
) -> Result<()> {
    let msg = plugin_config.to_supported_config_types_message()?;
    mqtt_client.published.send(msg).await
}
```

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 2]
```diff
@@ -134,73 +135,71 @@ async fn run(
...
+    loop {
+        tokio::select! {
+            message = mqtt_client.received.next() => {
+                if let Some(message) = message {
+                    debug!("Received {:?}", message);
+                    if let Ok(payload) = message.payload_str() {
+                            let result = match payload.split(',').next().unwrap_or_default() {
+                                "524" => {
+                                    let config_download_request =
+                                        SmartRestConfigDownloadRequest::from_smartrest(payload)?;
```
Comment (rina23q) [inline plugins/c8y_configuration_plugin/src/main.rs:154]: Why did you remove your own fix done by #1220? Now it crashes if you run the test case.

**Review by rina23q — COMMENTED** [review]
The `inotify` part works, but the changes you made killed your previous fix by #1220. It's an issue.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 3]
```diff
@@ -56,9 +57,6 @@ pub struct ConfigPluginOpt {
 
     #[clap(long = "config-dir", default_value = DEFAULT_TEDGE_CONFIG_PATH)]
     pub config_dir: PathBuf,
-
-    #[clap(long = "config-file", default_value = DEFAULT_PLUGIN_CONFIG_FILE_PATH)]
-    pub config_file: PathBuf,
```
[the remark this answers is not in the document]
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:61]: I wanted to remove it because in my opinion it is more confusing than helpful. Yes it can be nice to set the path to the configuration file to somewhere else, but then what is `--config-dir` doing anymore? 

For me `--config-dir` is the only place for configuration files. Having a `--config-file` can do more harm than good to the user. Also from our perspective it is more annoying to handle it.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 4]
```diff
@@ -134,73 +135,71 @@ async fn run(
...
-                        }
-                    }
-                    _ => {
-                        // Ignore operation messages not meant for this plugin
-                        Ok(())
+    let fs_notification_stream = fs_notify_stream(&[(
+        config_dir,
+        Some(config_file.to_string()),
+        &[FileEvent::Modified, FileEvent::Deleted, FileEvent::Created],
+    )])?;
```
[the remark this answers is not in the document]
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:140]: > * Is `config_file` equals to `c8y/c8y-configuration-plugin.toml` or `c8y-configuration-plugin.toml` ?
> * In other words, does the `fs_notify_stream` watch the directory recursively i.e including the sub-directories as `c8y`?

No it only watches for `c8y-configuration-plugin.toml` inside the c8y dir. But it supports a relative path to the file. Which is handled [here](https://github.com/thin-edge/thin-edge.io/blob/325935ef3c7c17d3c43e719211a410407a734cb3/crates/common/tedge_utils/src/fs_notify.rs#L141-L151).

[the remark this answers is not in the document]
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:140]: > Another comment about removing the `--config-file` option. Are you proposing to remove this option because of this call where you need a directory and some file to watch?

No only to keep the consistency across both log and configuration plugins. By design that helper function takes those arguments as inputs but it could have been abstracted in the API to take a full path to a file for a full path to a directory.

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply plugins/c8y_configuration_plugin/src/main.rs:61]: In my opinion, `--config-file` option is not good with our `fs_notify` crate. The crate needs to know the directory where the target file exists. 

However, user will give the path to the plugin configuration file by `--config-file`. e.g.
```
c8y_configuration_plugin --config-file '/etc/some/dir/plugin-config.toml
```
Note: with `--config-file` option, even the name of the configuration file has freedom to choose.
 
The config plugin needs to get the directory `/etc/some/dir` and the file name `plugin-config.toml` based on user input. Of course, it is possible to implement, but we need more efforts to support this.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 5]
```diff
@@ -56,9 +57,6 @@ pub struct ConfigPluginOpt {
 
     #[clap(long = "config-dir", default_value = DEFAULT_TEDGE_CONFIG_PATH)]
     pub config_dir: PathBuf,
-
-    #[clap(long = "config-file", default_value = DEFAULT_PLUGIN_CONFIG_FILE_PATH)]
-    pub config_file: PathBuf,
 }
 
 async fn create_mqtt_client(mqtt_port: u16) -> Result<mqtt_channel::Connection, anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (rina23q) [reply plugins/c8y_configuration_plugin/src/main.rs:62]: Whether the MQTT client subscribes the topic or not. Since we don't check the topic anymore, we should remove it. It confuses us later.

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply plugins/c8y_configuration_plugin/src/main.rs:61]: Regardless of which direction we will decide or not, the change will be user-faced change. We cannot change it without specification update if we want to change.

**Review by rina23q — COMMENTED** [review]
The code is very neat, and the regression was fixed. Now I see only one thing, **"whether we should remove `--config-file` option or not"**.

If yes,
* The specification must be updated accordingly.

If no,
* The code needs to be changed to address the case of `--config-file`.

Otherwise, this PR cannot get merged :/

**Comment by didier-wenzek** [comment]
> The code is very neat, and the regression was fixed. Now I see only one thing, **"whether we should remove `--config-file` option or not"**.
> 
> If yes,
> 
> * The specification must be updated accordingly.
> 
> If no,
> 
> * The code needs to be changed to address the case of `--config-file`.
> 
> Otherwise, this PR cannot get merged :/

Yes, the `--config-file` option can be removed and the documentation must be updated accordingly.

**Comment by didier-wenzek** [comment]
The clippy warning must be fixed too ;-)

**Comment by initard** [comment]
removed the `config-file` option and made some changes to the reference

**Review by rina23q — APPROVED** [review]
LGTM

File: docs/src/references/c8y-configuration-management.md [hunk 6]
```diff
@@ -120,19 +120,19 @@ USAGE:
...
         --config-dir <CONFIG_DIR>      [default: /etc/tedge]
-        --config-file <CONFIG_FILE>    [default: $CONFIG_DIR/c8y/c8y-configuration-plugin.toml]
         --debug                        Turn-on the debug log level
     -h, --help                         Print help information
     -i, --init                         Create supported operation files
     -V, --version                      Print version information
 
     On start, `c8y_configuration_plugin` notifies the cloud tenant of the managed configuration files,
-    listed in the `CONFIG_FILE`, sending this list with a `119` on `c8y/s/us`.
+    listed in the `CONFIG_DIR/c8y`, sending this list with a `119` on `c8y/s/us`.
```
Comment (didier-wenzek) [inline docs/src/references/c8y-configuration-management.md:129]:
```suggestion
    listed in `CONFIG_DIR/c8y/c8y-configuration-plugin.toml`, sending this list with a `119` on `c8y/s/us`.
```

File: docs/src/references/c8y-configuration-management.md [hunk 7]
```diff
@@ -120,19 +120,19 @@ USAGE:
...
     `c8y_configuration_plugin` subscribes then to `c8y/s/ds` listening for configuration operation
     requests (messages `524` and `526`).
     notifying the Cumulocity tenant of their progress (messages `501`, `502` and `503`).
     
     The thin-edge `CONFIG_DIR` is used to find where:
+    * to store the configuration file: c8y-configuration-plugin.toml
```
Comment (didier-wenzek) [inline docs/src/references/c8y-configuration-management.md:135]:
```suggestion
    * to store the configuration file: `c8y/c8y-configuration-plugin.toml`
```

**Review by didier-wenzek — COMMENTED** [review]
The code is correct but the documentation needs some changes.

**Review by didier-wenzek — APPROVED** [review]
Approved
